bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/firmware/FirmwareBuilder.java

```diff
@@ -0,0 +1,227 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.thing.binding.firmware;
+
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Map;
+import java.util.Objects;
+
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.internal.firmware.FirmwareImpl;
+
+/**
+ * The builder to create a {@link Firmware}.
+ *
+ * @author Thomas Höfer - Initial contribution
+ * @author Dimitar Ivanov - extracted as separate class for Firmware.
+ */
+public final class FirmwareBuilder {
+
+    private final ThingTypeUID thingTypeUID;
+    private final String version;
+    private String vendor;
+    private String model;
+    private boolean modelRestricted;
+    private String description;
+    private String prerequisiteVersion;
+    private String changelog;
+    private URL onlineChangelog;
+    private transient InputStream inputStream;
+    private String md5Hash;
+    private Map<String, String> properties;
+
+    public ThingTypeUID getThingTypeUID() {
+        return thingTypeUID;
+    }
+
+    public String getVersion() {
+        return version;
+    }
+
+    public String getVendor() {
+        return vendor;
+    }
+
+    public String getModel() {
+        return model;
+    }
+
+    public boolean isModelRestricted() {
+        return modelRestricted;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public String getPrerequisiteVersion() {
+        return prerequisiteVersion;
+    }
+
+    public String getChangelog() {
+        return changelog;
+    }
+
+    public URL getOnlineChangelog() {
+        return onlineChangelog;
+    }
+
+    public InputStream getInputStream() {
+        return inputStream;
+    }
+
+    public String getMd5Hash() {
+        return md5Hash;
+    }
+
+    public Map<String, String> getProperties() {
+        return properties;
+    }
+
+    /**
+     * Creates a new builder.
+     *
+     * @param thingTypeUID the thing type UID that is associated with this firmware (must not be null)
+     * @param firmwareVersion the version of the firmware to be created (must not be null)
+     * @throws NullPointerException if given thing type UID is null
+     * @throws IllegalArgumentException if given firmware version is null or empty
+     */
+    public FirmwareBuilder(ThingTypeUID thingTypeUID, String firmwareVersion) {
+        Objects.requireNonNull(thingTypeUID, "Thing type UID must not be null.");
```

Comment:
isn't it somewhat inconsistent to throw an NPE in case on argument is `null` and throw an IAE in case the other is `null`?

I won't repeat it, but this commend basically applies to the whole firmware area - it's pretty inconsistent in this respect. The rest of ESH avoids throwing NPEs "on purpose" - usually in such cases IAE are thrown in order to distinguish and "accidental programming bug" from a "hey, gotcha, you've been passing crap"-situation.

Reply:
PS: I of course know that it was like that previously - I just reckon this would be a good point in time to streamline this.